### ISSUE TITLE
Move retry on low level network errors to connection

### ISSUE THREAD

**Issue by tulios** [tulios/kafkajs#42]
Currently, all errors are bubbled up to high-level entities. KafkaJS should handle low-level errors such as `request timeout` inside the connection automatically.

Log example:
```
{
  broker: "kafka:9094",
  clientId: "myClientId",
  correlationId: 2785	
  error: "The request timed out",
  level: "ERROR",
  logger: "kafkajs",
  message: "[Connection] Response OffsetCommit(key: 8, version: 2)",
  size: 70	,
  timestamp: "2018-03-26T16:17:48.521Z"
}
```

**Comment by tulios** [comment]
@ashokkumarsand, I don't think your problem is related to this issue. Even if the consumer crashes it should reconnect after a while, from the error message it means that your `eachMessage` or `eachBatch` functions took longer than the session timeout, so the consumer was kicked-out of the group, thus the "The coordinator is not aware of this member" message.

You probably need a different configuration for the session timeout, take a look at: https://github.com/tulios/kafkajs#-options-1

**Comment by ashokkumarsand** [comment]
@tulios Thanks for the help. I will try

**Comment by HunderlineK** [comment]
@tulios Perhaps related to this. I've been simulating network failures by starting the consumer successfully and then disconnecting the network. This is the error I get

```
{"level":"ERROR","timestamp":"2020-06-17T17:51:39.366Z","logger":"kafkajs","message":"[Connection] Connection error: read ETIMEDOUT","broker":"url:9093","clientId":"kafka-consumer","stack":"Error: read ETIMEDOUT\n    at TLSWrap.onStreamRead (internal/stream_base_commons.js:205:27)"}
```

The problem is that even after enabling the network, the consumer does not resume. There are no 'CRASH' events from the consumer either.

**Comment by thomaspeklak** [comment]
@tulios I got the same error as @HunderlineK with kafkajs@1.14.0. The consumer stopped working. How do I recover from this error? 

Should 1.15 fix this problem?

**Comment by HunderlineK** [comment]
@thomaspeklak I implemented a custom retry mechanism to recover, which admittedly is not super clean, but it works:

There's a 'healthSignal' producer that regularly puts messages into a `healthSignal` topic on Kafka. All the consumers are subscribed to that topic in addition to the other topics they are normally subscribed too, so they should be consuming at least those healthSignal messages regularly even if there are no other messages coming into the topics they are subscribed too. Now, if a consumer does not consume any messages after a set interval, it will cause an alert and the node instance running kafkajs will automatically restart.

**Comment by thomaspeklak** [comment]
@HunderlineK This approach has a single point of failure: the healthSignal producer. If this goes down, all applications will permanently go into a reconnect or restart cycle (depending on your reaction to the failing health signal), meaning that all communication via kafka will halt. I am considering an approach similar to https://github.com/tulios/kafkajs/issues/452#issuecomment-517747429 . But there seems still room for improvement with this approach..

**Comment by HunderlineK** [comment]
@thomaspeklak Yeah, I see your point; it fit our architecture but will add a new point of failrue to others. The instance that consumes the messages is also the producer (it's one single instance than simply routes the messages to serverless functions for further processing), so having it didn't add a new point of failure.
